File: q_set2/prb1_2.java
public class EmployeeHierarchy {

    int findNextClosing(String stringToBeParsed, int startPoint) {
        int next = startPoint + 1;
        int inner = 0;
        while(next < stringToBeParsed.length()) {
            if (stringToBeParsed.charAt(next) == '(')
                inner += 1;
            if (stringToBeParsed.charAt(next) == ')' && inner != 0)
                inner -= 1;
            else if (stringToBeParsed.charAt(next) == ')' && inner == 0)
                return next;
            next += 1;
        }
        return -1;
    }

    int findNextOpening(String stringToBeParsed, int startPoint) {
        int next = startPoint + 1;
        if (stringToBeParsed.length() < 1) {
            return -1;
        } else {
            while (stringToBeParsed.charAt(next) != '(') {
                if (next == stringToBeParsed.length() - 1)
                    return -1;
                next += 1;
            }
            return next;
        }
    }

    void exterminateTheUnavailable(String stringToBeParsed){
        int openingParanthesis = 0;
        int closingParanthesis;
        System.out.println(stringToBeParsed.substring(0, stringToBeParsed.length()));

        while(openingParanthesis != -1) {
            closingParanthesis = findNextClosing(stringToBeParsed, openingParanthesis);

            if(stringToBeParsed.substring(openingParanthesis + 2, openingParanthesis + 13).equals("Unavailable")) {
                if (stringToBeParsed.substring(openingParanthesis, closingParanthesis).equals(stringToBeParsed)) {
                    stringToBeParsed = "";
                    break;
                }
                stringToBeParsed = stringToBeParsed.substring(0, openingParanthesis - 2) + stringToBeParsed.substring(closingParanthesis + 1);

            }
            openingParanthesis = findNextOpening(stringToBeParsed, openingParanthesis);
        }
        System.out.println(stringToBeParsed);
    }

    public static void main(String[] args) {
        EmployeeHierarchy employeeHierarchy = new EmployeeHierarchy();
        String inputString = "(“John”, (“Jasmine”, (“Jay”), (“Unavailable”)), (“Unavailable”, (“Jack”, (“Jeremy”))), (“Johanna”))";
        employeeHierarchy.exterminateTheUnavailable(inputString);
    }
}

File: q_set2/prb1_2.py
string = '("John", ("Jasmine", ("Jay"), ("Unavailable")), ("Unavailable", ("Jack", ("Jeremy"))), ("Johanna"))'


def find_next_closing(st: str, index: int):
    next = index + 1
    inner = 0
    while next < len(st):
        if st[next] == "(":
            inner += 1
        if st[next] == ")" and inner != 0:
            inner -= 1
        elif st[next] == ")" and inner == 0:
            return next
        next += 1


def find_next_opening(st: str, index: int):
    next = index + 1
    while st[next] != "(":
        if next == len(st) - 1:
            return -1
        next += 1
    return next


o = 0
c = 0
while o != -1:
    c = find_next_closing(string, o)
    if string[o + 1:o + 14] == '"Unavailable"':
        if string[o:c + 1] == string:
            string = ""
            break
        string = string[0:o - 2] + string[c + 1:len(string)]
    o = find_next_opening(string, o)

print(string)
